File: src/main/java/org/goafabric/dbagentnew/logic/PersonLogic.java
package org.goafabric.dbagentnew.logic;

import dev.langchain4j.agent.tool.Tool;
import org.goafabric.dbagentnew.persistence.PersonRepository;
import org.goafabric.dbagentnew.persistence.entity.Person;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.StreamSupport;

@Component
@Transactional
public class PersonLogic {

    private final PersonRepository personRepository;


    public PersonLogic(PersonRepository personRepository) {
        this.personRepository = personRepository;
    }

    public List<Person> findAll() {
        return StreamSupport.stream(personRepository.findAll().spliterator(), false)
                .toList();
    }


    @Tool
    public List<Person> findByFirstName(String firstName) {
        return personRepository.findByFirstNameIgnoreCase(firstName);
    }

    @Tool
    public List<Person> findByLastName(String lastName) {
        return personRepository.findByLastNameIgnoreCase(lastName);
    }


    @Tool
    public List<Person> findByCity(String city) {
        return personRepository.findByAddressCityContainsIgnoreCase(city);
    }

    public Person save(Person person) {
        return personRepository.save(person);
    }

}

File: src/main/kotlin/org/goafabric/llm/dbassistant/logic/PersonLogic.kt
package org.goafabric.llm.dbassistant.logic

import dev.langchain4j.agent.tool.Tool
import org.goafabric.llm.dbassistant.persistence.PersonRepository
import org.goafabric.llm.dbassistant.persistence.entity.Person
import org.springframework.stereotype.Component
import java.util.stream.StreamSupport

@Component //@Transactional
class PersonLogic(private val personRepository: PersonRepository) {
    fun findAll(): MutableList<Person> {
        return StreamSupport.stream<Person>(personRepository.findAll().spliterator(), false)
            .toList()
    }


    @Tool
    fun findByFirstName(firstName: String): MutableList<Person> {
        return personRepository.findByFirstNameIgnoreCase(firstName)
    }

    @Tool
    fun findByLastName(lastName: String): MutableList<Person> {
        return personRepository.findByLastNameIgnoreCase(lastName)
    }

    @Tool
    fun findByCity(city: String): MutableList<Person> {
        return personRepository.findByAddressCityContainsIgnoreCase(city)
    }

    @Tool
    fun findByAllergy(allergy: String): MutableList<Person> {
        return personRepository.findByAllergyAllergyContainsIgnoreCase(allergy)
    }

    fun save(person: Person): Person {
        return personRepository.save<Person>(person)
    }
}
